### PR TITLE
[GENX] Add `genx.matrix.2Dblockstore`

### PR DESCRIPTION
**Pull request by whitneywhtsang**
The `genx.matrix.2Dblockstore` operation stores to a submatrix from an array in memory.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 1]
```diff
@@ -441,6 +441,54 @@ def GENX_Matrix2DBlockLoadOp : GENX_Op<"matrix.2Dblockload">,
   let hasVerifier = 1;
 }
 
+def GENX_Matrix2DBlockStoreOp : GENX_Op<"matrix.2Dblockstore">,
+  Arguments<(ins 
+    Arg<LLVM_PointerTo<GENX_MatrixElemType>, "", [MemRead]>:$ptr,
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Isn't this written to ?

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp [hunk 2]
```diff
@@ -40,6 +40,22 @@ LogicalResult GENX::Matrix2DBlockLoadOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// genx.matrix.2Dblockstore
+//===----------------------------------------------------------------------===//
+
+LogicalResult GENX::Matrix2DBlockStoreOp::verify() {
+  if (getElemSizeInBits() != 8 && getElemSizeInBits() != 16 &&
```
Comment (etiotto) [inline mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:75]: Similar comment I made in another PR. Diagnostic should be more specific.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 3]
```diff
@@ -262,6 +262,44 @@ createGenISA2DBlockRead(GENX::Matrix2DBlockLoadOp op,
   return builder.CreateCall(fn, args);
 }
 
+// Create a call to GenISA_LSC2DBlockWrite for storing to a 2D submatrix.
+static void
+createGenISA2DBlockWrite(GENX::Matrix2DBlockStoreOp op,
+                         llvm::IRBuilderBase &builder,
+                         LLVM::ModuleTranslation &moduleTranslation) {
+  llvm::Module *module = builder.GetInsertBlock()->getModule();
+  auto opTypes = op->getOperandTypes();
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: Avoid using auto here.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 4]
```diff
@@ -262,6 +262,44 @@ createGenISA2DBlockRead(GENX::Matrix2DBlockLoadOp op,
...
+
+  // The IGC intrinsic requires the first argument be int64
+  assert(isa<llvm::PointerType>(args[0]->getType()) &&
+         "Expecting a pointer type");
+  args[0] = builder.CreatePointerCast(args[0], builder.getInt64Ty());
+  llvm::Value *lastOperand = args.pop_back_val();
+
+  llvm::ConstantInt *width = dyn_cast<llvm::ConstantInt>(args[1]);
+  llvm::ConstantInt *pitch = dyn_cast<llvm::ConstantInt>(args[2]);
+  assert((!width || !pitch || width->getZExtValue() <= pitch->getZExtValue()) &&
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: Please add this constraint in the verification code.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 5]
```diff
@@ -262,6 +262,44 @@ createGenISA2DBlockRead(GENX::Matrix2DBlockLoadOp op,
...
+
+  auto int32Ty = builder.getInt32Ty();
+  auto int1Ty = builder.getInt1Ty();
+  args.push_back(llvm::ConstantInt::get(int32Ty, op.getElemSizeInBits()));
+  args.push_back(llvm::ConstantInt::get(int32Ty, op.getTileWidth()));
+  args.push_back(llvm::ConstantInt::get(int32Ty, op.getTileHeight()));
+  args.push_back(llvm::ConstantInt::get(int32Ty, op.getVBlocks()));
+  args.push_back(llvm::ConstantInt::get(int1Ty, op.getTranspose()));
+  args.push_back(llvm::ConstantInt::get(int1Ty, op.getVnniTransform()));
+  args.push_back(lastOperand);
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:289]: why do we need to treat this in a special way ? Can we just pass `op.getStored_val()` ?

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir [hunk 6]
```diff
@@ -16,6 +16,22 @@ func.func @matrix_2Dblockload(%ptr : !llvm.ptr<i32>, %base_width : i32, %base_he
 
 // -----
 
+func.func @matrix_2Dblockstore(%ptr : !llvm.ptr<i32>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32, %stored_val : i32) {
+  // expected-error @+1 {{'genx.matrix.2Dblockstore' op invalid element size}}
```
Comment (etiotto) [inline mlir/test/Dialect/LLVMIR/genx-invalid.mlir]: Is not clear what is the error. Which operand has invalid element size ?

File: mlir/test/Dialect/LLVMIR/genx.mlir [hunk 7]
```diff
@@ -125,6 +125,12 @@ func.func @genx.2Dblockload1x4.32.1.0.0(%ptr : !llvm.ptr<i32>, %base_width : i32
   llvm.return
 }
 
+func.func @genx.2Dblockstore(%ptr : !llvm.ptr<i32>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32, %stored_val : i32) {
+  // CHECK: genx.matrix.2Dblockstore %arg0, %arg1, %arg2, %arg3, %arg4, %arg5, %arg6 {elem_size_in_bits = 32 : i32, tile_height = 1 : i32, tile_width = 4 : i32, transpose = false, v_blocks = 1 : i32, vnni_transform = false} : (!llvm.ptr<i32>, i32, i32, i32, i32, i32, i32)
```
Comment (etiotto) [inline mlir/test/Dialect/LLVMIR/genx.mlir]: Noticed that v_blocks is printed before vnni_transform, can you fix the print order pls.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 8]
```diff
@@ -441,6 +441,54 @@ def GENX_Matrix2DBlockLoadOp : GENX_Op<"matrix.2Dblockload">,
...
+    I32:$base_pitch,
+    I32:$x,
+    I32:$y,
+    I32Attr:$elem_size_in_bits,
+    I32Attr:$tile_width,
+    I32Attr:$tile_height,
+    I32Attr:$v_blocks,
+    I1Attr:$transpose,
+    I1Attr:$vnni_transform,
+    AnyInteger:$stored_val
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: This looks incorrect. How do we store fp16, bf16, fp32 ?

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 1 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: opps, fixed.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 8 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: According to https://github.com/intel-innersource/drivers.gpu.unified/blob/comp/igc/Source/IGC/GenISAIntrinsics/Intrinsic_definitions.py#L2283, stored_val should be any int.

File: mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply mlir/lib/Dialect/LLVMIR/IR/GENXOps.cpp:75]: done.

File: mlir/test/Dialect/LLVMIR/genx.mlir (hunk 7 above)
Comment (whitneywhtsang) [reply mlir/test/Dialect/LLVMIR/genx.mlir]: It is printed in alphabetical order.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 8 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: I think the operation we design needs to store a block of elements so IMO this operand should have type vector< GENX_MatrixElemType>. I should be able to load a block or element and store it back using this operation.

Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: The hardware 2D load/store instructions are type-less (just the bits) so IGC intrinsics defines the store value as anyint.  We will need to bitcast the vector of fp16, bf16, fp32, int8 to vector of int32 when we translate this to IGC function call, so we can make the $stored_val to be any FixedVectorOf<[GENX_MatrixElemType]>.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 9]
```diff
@@ -441,6 +441,54 @@ def GENX_Matrix2DBlockLoadOp : GENX_Op<"matrix.2Dblockload">,
...
+  string baseDescription = [{
+    The 'genx.matrix.2Dblockstore' operation stores to a submatrix from an array in memory.
+    $ptr - the base address of the memory array
+    $base_width, $base_height, $base_pitch - the shape of the memory array
+    $x, $y, $tile_width, $tile_height - the starting offets and shape of the submatrix to load
+    $elem_size_in_bits - 32 for f32, bf32; 16 for f16, int16, bf16; 8 for int8, int4, int2 and etc
+    $v_blocks - number of blocks to load
+    $transpose - transpose the submatrix in vector register (useful for 32 bit element types)
+    $vnni_transform - transpose and pack the submatrix in register (useful for < 32 bit element types)
+    $stored_val - value to store
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: block to store ?

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 8 above)
Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: I think that the store value vector shape doesn't change:  vector<8xfp32> cast to vector<8xint32>, vector<8xbf16> cast to vector<8xint16>, vector<8xbf16> cast to vector<8xint16>, and etc. should work.

Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: @pengtu How can I store a vector of int32 to an integer? 
The IGC intrinsic expects anyint, and we can pass a vector?

**Review by pengtu — APPROVED** [review]

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 8 above)
Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: @whitneywhtsang: I think that we can pass int32 or int16 type as store the value type to llvm::GenISAIntrinsic::getDeclaration call so IGC will get the right function, and pass the vector store value to the function call.

Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: This way, we don't even need to bitcast to int.

Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: ok, so `GenISA_LSC2DBlockWrite` accept vector of any element type as store value, and we always pass `int32` to `getDeclaration` to get the  right function?

Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: With the latest clarification from IGC, we can pass vector of {int32, int16, int8} to "anyint" based on the original vector value's element type.

Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Thanks for clarifying, what should we do when element type is fp? Should I use the int type with the same size?

Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Given a store value of vector<totalSize x elemType>, we create a vector type of <totalSize/16 x intElemSize_t> as the store value type to getDeclaration call.

**Review by pengtu — APPROVED** [review]

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 8 above)
Comment (pengtu) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: A further clarification on this SIMT function that applies to a work item (simd lane): each work item supplies its own part of the submatrix values, i.e., one row of the block.  For PVC SIMD 16, if the submatrix block is an array of "float a[16, 4]", the stored value will be a[workitem_id, 0:3] -- a vector<4xf32>. 

IGC team has confirmed this.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: done.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 5 above)
Comment (whitneywhtsang) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:289]: We need the converted operand, `op.getStoredVal()` is not translated.

File: mlir/test/Dialect/LLVMIR/genx-invalid.mlir (hunk 6 above)
Comment (whitneywhtsang) [reply mlir/test/Dialect/LLVMIR/genx-invalid.mlir]: changed.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp]: done.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 9 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: done.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 10]
```diff
@@ -441,6 +441,54 @@ def GENX_Matrix2DBlockLoadOp : GENX_Op<"matrix.2Dblockload">,
...
+
+  let summary = "GENX 2D block store";
+
+  string baseDescription = [{
+    The 'genx.matrix.2Dblockstore' operation stores to a submatrix from an array in memory.
+    $ptr - the base address of the memory array
+    $base_width, $base_height, $base_pitch - the shape of the memory array
+    $x, $y, $tile_width, $tile_height - the starting offets and shape of the submatrix to load
+    $elem_size_in_bits - 32 for f32, bf32; 16 for f16, int16, bf16; 8 for int8, int4, int2 and etc
+    $v_blocks - number of blocks to load
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: to store ?

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 5 above)
Comment (etiotto) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:289]: ok

**Review by etiotto — APPROVED** [review]

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 10 above)
Comment (whitneywhtsang) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: done.

**Pull request merged**